### ISSUE TITLE
feature: make menu-bar keyboard navigable

### ISSUE THREAD

**Issue by arxanas** [arxanas/scm-record#44]
Typical GUIs let you open menu bar items using the keyboard. It would be nice if the menu bar items could be at least expanded using the keyboard.

This was originally requested on Discord: https://discord.com/channels/968932220549103686/1235711594185621585

**Comment by emesterhazy** [comment]
Yeah, this would be great. I used scm-record for a few months before I realized that I could click on the menu bars to learn the keybindings and do things.

We might want to show the keyboard shortcut in parentheses next to the menu item to make it clear how to expand it with the keyboard.

Perhaps we could just use numbers starting from 1? For example:

```
[File (1)] [Edit (2)] [Select (3)] [View (4)]
```

**Comment by arxanas** [comment]
> Perhaps we could just use numbers starting from 1? For example:

My dream would be to one day support multi-way splits where you could assign a change to the nth split by pressing the corresponding number key 😍. Perhaps we could use the F-keys for the menu items, similarly to GUI applications?

**Comment by emesterhazy** [comment]
That should work too :)

**Comment by essiene** [comment]
we could take a page from `nano`[1]. It has a footer palette that shows the most important keyboard shortcuts. We could have something for just the main menu items and the rest will be visible once the menu is activated.

> [File (1)] [Edit (2)] [Select (3)] [View (4)]

fwiw, I'm more used to seeing Menu type interfaces use either CTRL+ or MOD+ for these, instead of grabbing the number keys. So `CTRL+F` or 'MOD+F' to open the file menu would seem less surprising for me, plus I may be able to guess it without referring to docs if 'F' were in bold or underlined, or had some other hint.

> Perhaps we could use the F-keys for the menu items, similarly to GUI applications?

I'm personally not familiar with GUI apps that use F-keys for the menu? Though I have to confess my GUI apps these days is really just a browser. 

Also, some keyboards make the F-keys a pain to access via secondary combos (or don't even label them... I'm looking at your Chromebooks).

[1] - https://www.nano-editor.org/

**Comment by WhyNotHugo** [comment]
The standard shortcut for accessing menubars for the last several decades is <kbd>Alt</kbd>. Opening File would be <kbd>Alt</kbd>+<kbd>f</kbd>. I think it's best to stick to the standard. I honestly don't think I've ever seen any exception since the days of MS-DOS until today.

**Comment by WhyNotHugo** [comment]
The <kbd>Alt</kbd>+<kbd>letter</kbd> style is for DOS, Windows, and Linux. Like I said, this combination is pretty ubiquitous, I don't think I've ever seen applications with a menubar and a different shortcut to access it.

I haven't used windows in a while (I think that last one I actively used was Windows 7), but generally all the windows built-in apps follows this pattern (e.g.: notepad, paint, internet explorer, file explorer, etc). MS Office had the same. A quick online search indicates that Adobe software does the same.

GTK2 and GTK3 applications followed the same pattern (GTK4 has no menusbars). Qt applications too (I just double checked with a few, including mumble).

Finally, even Firefox will now show the menubar if you press <kbd>Alt</kbd>, and will poen the File menu if I press <kbd>Alt</kbd>+<kbd>f</kbd>

**Comment by WhyNotHugo** [comment]
macOS doesn't open menus with <kbd>Alt</kbd>. macOS has no <kbd>Alt</kbd> key. A quick online search indicates that <kbd>Control</kbd>+<kbd>F2</kbd> will focus the menubar. But note that macOS uses an _external_ menubar, so you don't want to override this shortcut (nor can you).

**Comment by WhyNotHugo** [comment]
> Ultimately: I don't think the menu bar is even an ideal system:

It's possible that it isn't. A menubar is mostly idea if you have a large amount of actions that are grouped in a way that aligns with it, and need a way for them to all be discoverable. It's only a good fit if you expect to eventually have more and more actions.

I think a searchable command pallette is not a good fit. It is good interface when you have **many** actions, and the user knows what they're called or how to look for them (or when actions are user-created). Actions are not discoverable at all; the user needs to know what they're called ahead of time.

If there will only be 7 to 14 actions, using a bar at the bottom like nano is not a bad idea; all shortcuts can be shown on screen at all times. But if the amount of actions ever grows beyond that, you start to need three rows of actions which is a bit much. Of course, a second layer of actions can be hidden behind a <kbd>?</kbd> mapping which shows a screen with their shortcuts -- this has the downside of them not being mouse-accessible at all, but I can't imagine power users using advanced actions complaining about this.

**Comment by tschundler** [comment]
> a searchable command pallette is not a good fit
Agreed

I like the either of their a top bar like `hg`'s `crecord` tool, with a `?` for more, or `nano` style common keys listed at the bottom. (personally I like the way `zellij` presents its short cuts on the bottom)

> this has the downside of them not being mouse-accessible at all
The only use case I can image where I might prefer "mouse" interaction is on a phone or tablet to get the keyboard out of the way and see more. But even then, the keyboard is always still an option. Are there other use cases for mouse/pointer/menu type interaction?
